zerver/lib/bugdown/__init__.py

```diff
@@ -1785,6 +1785,14 @@ def get_full_name_info(realm_id: int, full_names: Set[str]) -> Dict[str, FullNam
     if not full_names:
         return dict()
 
+    # Remove the trailing part of the `user|id` mention syntax.
+    name_re = r'(.+)\|\d+$'
+    for full_name in full_names.copy():
+        name_syntax_match = re.match(name_re, full_name)
+        if name_syntax_match:
+            full_names.remove(full_name)
+            full_names.add(name_syntax_match.group(1))
+
```

Comment:
Can you explain why this commit is correct?

Reply:
(I guess one thing that makes this commit hard to evaluate is that I'm not 100% sure without re-reading the code what this data structure is for; is it just the mention data for users potentially mentioned in this message, or for all users in the organization?)  And the other is that you don't have tests here for corner cases around duplicate users.  

But I wouldn't have expected there to be any special code for "duplicate names" on the backend; we're just adding support for a new syntax, here.

Reply:
Ah I should have added some more tests here. We initially populate the possible users that could have been mentioned in the message, and then bugdown uses this data structure to query them.

For initial fetch from the database, we pass all the user mentions and fetch exact full name matches from the database. Let's simply run through with an example:

1. Message: `@User1|123 and @User2 are mentioned.`
2. Passed to get_full_name_info: `Set('User1|123', 'User2')`
3. My Code (where previously was nothing)
4. Calls to database needs to be: `Set('User1', 'User2')` which will return all matches of User1 and User2. 

So what my code here is doing is looping through the (copy of the) Set, and replacing (removing, reinserting) `User1|123 -> User1`.

Reply:
I guess the logic that I'd had in mind is that we should be just looking up the user by ID, and then checking if the name matches (if we think that's important), but the lookup process should just ignore the full name part, since that's a cleaner architecture.

Reply:
We'd still need to support the full name lookup anyway, but if the id is present then we can lookup by id and thus do both forms of lookup. Unless we plan to enforce compulsory user id lookup, I do not think we should have both of them present.